Separate `oled_task_user()` from `drivers/oled/oled_driver.c`.

mtei opened the pull request:
## Description

Separate `oled_task_user()` from `drivers/oled/oled_driver.c` into a library that is linked as needed.

**This change is required by #12041.**

Delete the following line at the end of `drivers/oled/oled_driver.c`. ~~Because `oled_task_user()` is a function that the user must implement, not an option. If `oled_task_user()` is not defined, the build must fail.~~

```c
__attribute__((weak)) void oled_task_user(void) {}
```

And add drivers/oled/oled_driver_user_task.c that have empty `oled_task_user()`. This will be a library as shown below and will be linked if the user did not define `oled_task_user()`.

```makefile
// $(TOP)/common_features.mk
ifeq ($(strip $(OLED_DRIVER_ENABLE)), yes)
    OPT_DEFS += -DOLED_DRIVER_ENABLE
    COMMON_VPATH += $(DRIVER_PATH)/oled
    QUANTUM_LIB_SRC += i2c_master.c
    QUANTUM_LIB_SRC += oled_driver_user_task.c
    SRC += oled_driver.c
endif
```

This change provides more ways to provide the keyboard-level default oled_task_user(). See https://github.com/qmk/qmk_firmware/pull/12113#issuecomment-791944270 and https://github.com/qmk/qmk_firmware/pull/12041#issuecomment-788925669 for details.


Tagging: @XScorpion2 

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

XScorpion2 commented:
>  Because oled_task_user() is a function that the user must implement, not an option. If oled_task_user() is not defined, the build must fail.

I can understand that the feeling is this function is not optional, but I don't fully agree with this statement. It is entirely possible to write to the oled entirely from other qmk functions or events bypassing the oled_task_user() function. The benefit that oled_task_user() function provides over driving it via events in other functions is that it is a central entry point where the cursor is normalized to the beginning of the display and consistent call timing. Depending on your approach and desired display setup, in many cases it's can be better to write content as events occur and control the cursor manually, thus with this change be required to implement an empty oled task function.

mtei commented:
Okay, I understand that there are cases where you need an empty `oled_task_user()`.
Fixed to provide it in a different way than `__attribute__ ((weak))`.

see https://github.com/qmk/qmk_firmware/pull/12113/commits/814d61ce12fc27fdb43382c6b6a8c9f99a29d13c

zvecr commented on drivers/oled/oled_driver_user_task.c:
```diff
@@ -0,0 +1,3 @@
+#include "oled_driver.h"
+
+void oled_task_user(void) { }
```
As this is always added to `QUANTUM_LIB_SRC`, this wont work without it being declared weak?

mtei replied:
```makefile
ifeq ($(strip $(OLED_DRIVER_ENABLE)), yes)
    OPT_DEFS += -DOLED_DRIVER_ENABLE
    COMMON_VPATH += $(DRIVER_PATH)/oled
    QUANTUM_LIB_SRC += i2c_master.c
    QUANTUM_LIB_SRC += oled_driver_user_task.c
    SRC += oled_driver.c
endif
```

If `oled_task_user()` is defined in the keyboard or keymap code, `oled_driver_user_task.o` will not be linked. That is the function of QUANTUM_LIB_SRC and LIB_SRC.

zvecr replied:
Given we use weak for virtually every single point of customisation, i really dont see the need to fragment with a different way to do things.

mtei replied:
There is no change in the existing customization methods for users. It will still be overridden if the user defines oled_task_user().

This PR also provides a way to override the default oled_task_user() and other functions in a keyboard-level file with a keymap-level implementation of oled_task_user().

mtei replied:
There is no change in the existing customization methods for users. It will still be overridden if the user defines oled_task_user().

This PR also provides a way to override the default oled_task_user() and other functions in a keyboard-level file with a keymap-level implementation of oled_task_user().

zvecr reviewed (commented):
Also, changed base to develop, as the PRs intent is to change an existing behaviour.

mtei commented:
> Also, changed base to develop, as the PRs intent is to change an existing behaviour.

This PR does not change the behavior of the resulting firmware. So I don't think we need to change it to develop.

In the '0.12.11', there are 249 patterns that declare `OLED_DRIVER_ENABLE = yes`. Of these, 219 patterns can be built normally.

<details>
<summary><strong>test script (click)</strong></summary>

```shell
# LOG=/tmp/before-md5
LOG=/tmp/after-md5
export SKIP_VERSION=yes
rm -f $LOG
make clean
make SILENT=true 10bleoledhub:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true 10bleoledhub:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true anavi/macropad8:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true anavi/macropad8:kicad:check-md5 | grep ^MD5 >> $LOG
make SILENT=true anavi/macropad8:kodi:check-md5 | grep ^MD5 >> $LOG
make SILENT=true anavi/macropad8:obs:check-md5 | grep ^MD5 >> $LOG
make SILENT=true anavi/macropad8:zoom:check-md5 | grep ^MD5 >> $LOG
make SILENT=true angel64/alpha:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true angel64/rev1:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true angel64/rev1:kakunpc:check-md5 | grep ^MD5 >> $LOG
make SILENT=true aplyard/aplx6/rev2:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true arabica37/rev1:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true arch_36:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true arch_36:default_split_3x5_3:check-md5 | grep ^MD5 >> $LOG
make SILENT=true arch_36:obosob:check-md5 | grep ^MD5 >> $LOG
make SILENT=true basekeys/slice/rev1_rgb:2moons_rgb:check-md5 | grep ^MD5 >> $LOG
make SILENT=true basekeys/slice/rev1_rgb:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true basekeys/slice/rev1_rgb:default_split_left_space:check-md5 | grep ^MD5 >> $LOG
make SILENT=true basekeys/slice/rev1_rgb:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true boardsource/holiday/spooky:rip_mx:check-md5 | grep ^MD5 >> $LOG
make SILENT=true boardsource/holiday/spooky:rip_my_wallet:check-md5 | grep ^MD5 >> $LOG
make SILENT=true boardsource/microdox:cole:check-md5 | grep ^MD5 >> $LOG
make SILENT=true boardsource/microdox:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true business_card/alpha:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true business_card/beta:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true chidori:oled_sample:check-md5 | grep ^MD5 >> $LOG
make SILENT=true ckeys/washington:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true claw44/rev1:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/common:curry:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/common:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/common:drashna:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/common:gotham:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/common:kidbrazil:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/common:ninjonas:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/common:oled_sample:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/common:rjhilgefort:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/common:rpbaptist:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/common:soundmonster:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/common:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/common:xyverz:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/legacy:curry:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/legacy:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/legacy:gotham:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/legacy:kidbrazil:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/legacy:ninjonas:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/legacy:oled_sample:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/legacy:rjhilgefort:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/legacy:rpbaptist:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/legacy:soundmonster:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/legacy:xyverz:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/rgb:curry:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/rgb:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/rgb:gotham:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/rgb:kidbrazil:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/rgb:ninjonas:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/rgb:oled_sample:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/rgb:rjhilgefort:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/rgb:rpbaptist:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/rgb:soundmonster:check-md5 | grep ^MD5 >> $LOG
make SILENT=true crkbd/rev1/rgb:xyverz:check-md5 | grep ^MD5 >> $LOG
make SILENT=true dekunukem/duckypad:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true dekunukem/duckypad:m4cs:check-md5 | grep ^MD5 >> $LOG
make SILENT=true dekunukem/duckypad:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true dmqdesign/spin:gorbachev:check-md5 | grep ^MD5 >> $LOG
make SILENT=true doodboard/duckboard:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true dumbo:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true dumbo:trip-trap:check-md5 | grep ^MD5 >> $LOG
make SILENT=true gergo:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true getta25/rev1:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/d48:anderson:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/d48:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/dactyl_manuform/5x6_5:333fred:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/dactyl_manuform/5x6_right_trackball:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/dactyl_manuform/5x6_right_trackball:drashna:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/myskeeb:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/obuwunkunubi/spaget:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/onekey/blackpill_f401:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/onekey/blackpill_f411:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/onekey/bluepill:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/onekey/elite_c:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/onekey/promicro:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/onekey/proton_c:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/onekey/stm32f0_disco:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/onekey/teensy_2:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/onekey/teensy_2pp:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/onekey/teensy_32:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/onekey/teensy_lc:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/owlet60:oled_testing:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/pill60/blackpill_f401:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/pill60/blackpill_f411:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/pill60/bluepill:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/riblee_f411:333fred:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/riblee_f411:alfrdmalr:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/riblee_f411:brandonschlack:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/riblee_f411:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/riblee_f411:default_ortho_5x12:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/riblee_f411:drashna:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/riblee_f411:greatwizard:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/riblee_f411:peej:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/riblee_f411:riblee:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/riblee_f411:rs:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/riblee_f411:xyverz:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/swiftrax/koalafications:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/swiftrax/koalafications:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/swiftrax/the_galleon:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true handwired/swiftrax/the_galleon:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev2:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev2:led_test:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev2:xulkal:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev2/back:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev2/back:led_test:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev2/back:xulkal:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev2/qmk_conf:xulkal:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev2/sc:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev2/sc:led_test:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev2/sc:xulkal:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev2/under:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev2/under:led_test:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev2/under:xulkal:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev3_4rows:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev3_4rows:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev3_5rows:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev3_5rows:five_rows:check-md5 | grep ^MD5 >> $LOG
make SILENT=true helix/rev3_5rows:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true jagdpietr/drakon:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true jagdpietr/drakon:wkl:check-md5 | grep ^MD5 >> $LOG
make SILENT=true keybage/radpad:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true keycapsss/kimiko/rev1:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kikoslab/kl90:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kikoslab/kl90:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kingly_keys/romac:boss566y:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kingly_keys/romac_plus:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kingly_keys/romac_plus:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true knobgoblin:ortho:check-md5 | grep ^MD5 >> $LOG
make SILENT=true knobgoblin:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:asapjockey:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:benji:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:cjuniet:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:corodiak:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:cwebster2:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:drashna:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:ghidalgo93:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:gotham:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:j-inc:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:jhelvy:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:mattir:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:ninjonas:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:plattfot:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:shinze:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:tessachka:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:thomasbaart:check-md5 | grep ^MD5 >> $LOG
make SILENT=true kyria/rev1:winternebs:check-md5 | grep ^MD5 >> $LOG
make SILENT=true latinpad:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true latinpad:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true latinpadble:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true latinpadble:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lck75:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:barabas:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:bcat:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:chuan:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:curry:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:cykedev:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:domnantas:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:drasbeck:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:hvp:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:jhelvy:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:lily58l:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:mikefightsbears:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:muuko:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:ninjonas:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:yshrsmz:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/light:yuchi:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:barabas:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:bcat:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:chuan:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:curry:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:cykedev:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:domnantas:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:drasbeck:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:hvp:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:jhelvy:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:lily58l:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:mikefightsbears:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:muuko:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:ninjonas:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:yshrsmz:check-md5 | grep ^MD5 >> $LOG
make SILENT=true lily58/rev1:yuchi:check-md5 | grep ^MD5 >> $LOG
make SILENT=true mechllama/g35/v1:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true mechllama/g35/v2:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true mechwild/mercutio:bongocat:check-md5 | grep ^MD5 >> $LOG
make SILENT=true mechwild/mercutio:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true mechwild/mercutio:fancy:check-md5 | grep ^MD5 >> $LOG
make SILENT=true mechwild/mercutio:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true nullbitsco/nibble:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true nullbitsco/nibble:oled_bongocat:check-md5 | grep ^MD5 >> $LOG
make SILENT=true nullbitsco/scramble:oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true palette1202:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true palette1202:key-check:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rainkeeb:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rainkeeb:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true ramonimbao/herringbone/pro:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true ramonimbao/herringbone/pro:iso:check-md5 | grep ^MD5 >> $LOG
make SILENT=true ramonimbao/herringbone/pro:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rart/rartand:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rart/rartand:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rart/rartlice:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rart/rartlice:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rgbkb/pan/rev1/32a:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rgbkb/pan/rev1/proton_c:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rgbkb/sol/rev1:brianweyer:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rgbkb/sol/rev1:xulkal:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rgbkb/sol/rev1:xyverz:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rgbkb/sol/rev2:brianweyer:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rgbkb/sol/rev2:xulkal:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rgbkb/sol/rev2:xyverz:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rgbkb/zygomorph/rev1:default_oled:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rocketboard_16:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true rocketboard_16:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true sendyyeah/pix:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true sendyyeah/pix:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true setta21/rev1:salicylic:check-md5 | grep ^MD5 >> $LOG
make SILENT=true sofle/rev1:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true sofle/rev1:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true spacetime/rev2:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true spacetime/rev2:kyleterry:check-md5 | grep ^MD5 >> $LOG
make SILENT=true splitkb/zima:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true splitkb/zima:drashna:check-md5 | grep ^MD5 >> $LOG
make SILENT=true suihankey/alpha:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true suihankey/rev1:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true tkc/m0lly:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true tkc/m0lly:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true tkc/tkc1800:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true tkc/tkc1800:smt:check-md5 | grep ^MD5 >> $LOG
make SILENT=true tkc/tkc1800:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true tkc/tkc1800:wkl:check-md5 | grep ^MD5 >> $LOG
make SILENT=true torn:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true torn:kinesish:check-md5 | grep ^MD5 >> $LOG
make SILENT=true torn:via:check-md5 | grep ^MD5 >> $LOG
make SILENT=true treadstone48/rev1:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true treadstone48/rev1:like_jis:check-md5 | grep ^MD5 >> $LOG
make SILENT=true treadstone48/rev1:like_jis_rs:check-md5 | grep ^MD5 >> $LOG
make SILENT=true treadstone48/rev2:default:check-md5 | grep ^MD5 >> $LOG
make SILENT=true treadstone48/rev2:like_jis:check-md5 | grep ^MD5 >> $LOG
make SILENT=true ungodly/launch_pad:default:check-md5 | grep ^MD5 >> $LOG
```
</details>

The build results are unchanged except for the following 16 patterns of binaries. And the following 16 patterns do not change the behavior of the binary.

* `make SILENT=true ungodly/launch_pad:default`

  This pattern will change the address of `oled_task_user()`, but not the behavior, because it has the following characteristics:

  * oled_init_user() defined in ungodly/launch_pad/keymaps/default/keymap.c
  * oled_task_user() undefined in ungodly/launch_pad/keymaps/default/keymap.c
  * oled_write_raw_P() called from ungodly/launch_pad/keymaps/default/keymap.c

* Remaining 15 patterns

  The following pattern does not use oled_driver in effect.

  * oled_init_user() undefined
  * oled_task_user() undefined
  * no call any functions in drivers/oled/oled_driver.c from keyboard and keymap codes

  Therefore, although OLED_DRIVER_ENABLE is declared as yes, they do not use drivers/oled/oled_driver.c.

  ```shell
  make SILENT=true arch_36:default_split_3x5_3
  make SILENT=true handwired/riblee_f411:333fred  
  make SILENT=true handwired/riblee_f411:alfrdmalr  
  make SILENT=true handwired/riblee_f411:default  
  make SILENT=true handwired/riblee_f411:default_ortho_5x12  
  make SILENT=true handwired/riblee_f411:greatwizard  
  make SILENT=true handwired/riblee_f411:peej  
  make SILENT=true handwired/riblee_f411:rs  
  make SILENT=true handwired/riblee_f411:xyverz  
  make SILENT=true kingly_keys/romac_plus:via
  make SILENT=true kyria/rev1:corodiak
  make SILENT=true lily58/light:bcat
  make SILENT=true lily58/rev1:bcat  
  make SILENT=true spacetime/rev2:default
  make SILENT=true spacetime/rev2:kyleterry
  ```

mtei commented:
> However if the intent is to go with 814d61c, i'm not sure i see the value in the current changeset. What does the extra compilation unit of a single function add?

At the keyboard level, you may want to have a default OLED code as shown below, and not link the default code once the keymap has defined its own unique OLED code.
We don't want to override just one 'oled_task_user()' function, we want to override one file.

```makefile
# keyboard level rules.mk
   LIB_SRC += default_oled_task_user_and_other_OLED_codes.c
```

If there is  `__attribute__((weak)) void oled_task_user(void)` in `driver/oled/oled_driver.c`, the linker will not link the default file above because `oled_task_user()` will always be defined.

This is why we need `QUANTUM_LIB_SRC += oled_driver_user_task.c`.
(See also here: https://github.com/qmk/qmk_firmware/pull/12041#issuecomment-788925669)

zvecr commented:
> This PR does not change the behaviour of the resulting firmware. So I don't think we need to change it to develop.

The key point here is "intent", and the fact your PR still states "that the user must implement". That change would have to go via develop as it is a change in an existing behaviour.

The behaviour of multiple weak of the same symbol is being used elsewhere. eg <https://github.com/qmk/qmk_firmware/blob/master/keyboards/splitkb/zima/zima.c#L40> and works as expected with user keymaps. While this arguably might produce a more well defined behaviour, I still think the change is totally redundant given the rest of the code base.

mtei commented:
> The key point here is "intent", and the fact your PR still states "that the user must implement".  

 Okay, I understand what you mean by "change an existing behaviour". It was my misunderstanding that "oled_task_user() must be implemented". That was revealed by the comments of XScorpion2. And I found that `ungodly/launch_pad:default` is an example. So I fixed the documentation.

mtei commented:
> Also, changed base to develop, as the PRs intent is to change an existing behaviour.

I've fixed it so that there is no change in behavior, so changed base to master.

> I still think the change is totally redundant given the rest of the code base.

This change is necessary, please see the description I added.

mtei commented:
Using the library to provide a default implementation is a very traditional practice. I don't think it's more common to have a default implementation in a weak function.

As explained above, if you provide a default implementation of the `oled_task_user()` function as a weak function in oled_display.c, you will not be able to provide a default implementation by the library at the keyboard level.  Also, as explained in #12041, helix/rev3_Xrows currently should override one file instead of one function.  Therefore, this PR must be implemented for the #12041 implementation (LIB_SRC = oled_display.c) to work.

drashna commented:
I meant to reply sooner, sorry.

> Using the library to provide a default implementation is a very traditional practice. I don't think it's more common to have a default implementation in a weak function.

I can't comment on if that is standard practice or not, since most of my experience is directly QMK related. 

However, this implementation is a significant departure from how QMK Firmware is set up, and it's practices.  I don't know of one instance where a file is replaced by a user file, ... well, outside of uGFX.   In every other case (eg, pretty much all of QMK Firmware), the standard way is by using weak functions, or boolean functions.  And I don't think that it's a good idea to start changing this here. 

To be blunt, I would prefer that `oled_display.c` was removed, and the code for the helix moved to the keyboard files instead.  Though, replacing the functions with weak versions is fine, based on how things are set up, already.

And to be absolutely clear, this is not a criticism on you or your skill (as both you and your skills are absolutely amazing, IMO), just that it doesn't fall in line with how things have been done in QMK firmware.
